### ISSUE TITLE
Security Requirements for WebHook Consumer

### ISSUE THREAD

**Issue by egekorkan** [w3c/wot-profile#222]
Given that a WebHook consumer also acts as a server, what are its security scheme requirements? If there is any flexibility on this, the Consumer needs to communicate these with a certain format, maybe a TD with no affordance but only security information?

**Comment by egekorkan** [comment]
In the call of 01.02.2022, there was a mention that the Consumer would have the security of the Thing. I don't think this is realistic given that a Consumer can be a cloud service and the Thing is something in the local network. They can use the same one but making it mandatory that they are the same seems a bit off to me. Do we any examples?

**Comment by mmccool** [comment]
Will discuss in next Security TF call.

**Comment by benfrancis** [comment]
> I (the Security TF is still reviewing and will comment here) suggests that we could adopt the approach outlined for Cloud Events in https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/http-webhook.md, with some modifications

I agree that using Bearer tokens in an Authorization header in event requests sent from a Thing to a Consumer is a reasonable solution.

Some observations:
- I note that this document doesn't deal with the subscription mechanism (i.e. how the token is provided to the "client" (Thing) by the "delivery target" (Consumer)), so a payload would need designing for the subscribe operation
- The document also doesn't say what to do when a token expires, which may occur between a subscribe and unsubscribe operation. Presumably the Thing would receive a 401 response from the Consumer when trying to deliver an event and would need to re-authenticate somehow (e.g. using a refresh token).
- It's notable that there's no way to describe Thing to Consumer authentication like this in a Thing Description, this is another area where the profile would have to go beyond what can be described in a Thing Description alone. This kind of thing could potentially [be defined as a default in a Webhook protocol binding](https://github.com/w3c/wot-binding-templates/issues/248), but defining a vocabulary to override the default would be tricky since the metadata would need to describe the Consumer, not the Thing.

**Comment by benfrancis** [comment]
Another source of inspiration could be the [W3C WebSub](https://www.w3.org/TR/websub) specification. That specification is a little different because in WebSub all Webhooks are managed by a "hub" which acts as an intermediary between publishers and subscribers. Such hubs may have more resources than some Things and therefore may be able to process more complex cryptography.

Anyway, the approach they use for "[authenticated content distribution](https://www.w3.org/TR/websub/#signing-content)" is for a subscriber to supply a "secret" in their subscription request which is then used by the hub to create an HMAC signature of each event payload, which is then included in a request header of each event notification. Subscribers can verify event notifications by using the shared secret to recompute the signature for a given payload and check that it matches.

The signing of individual payloads would provide an additional layer of protection when compared with the simple Bearer authentication token approach described above (by guarding against man-in-the-middle attacks), but would also be more computationally expensive.

Perhaps signing of payloads could be explored as an optional layer of security at the same time as signing Thing Descriptions.
